### DISCUSSION TITLE
Bad performance for 2 crazyflies and one radio

### DISCUSSION THREAD

**Discussion by Allanadn** [bitcraze/discussions#1041] [Q&A]
Dear Bitcraze team,

I am trying to fly two CFs using one radio. I am just trying some very simple commands using the swarm interface. Nonetheless, while one of the UAVs take off and does not goto the desired position, the other one goes completely nuts. As soon as it takes off (to a completely different direction), it keeps flying until it smash itself into something. I am using the Flowdeck v2 on both UAVs and I have absolutely no clue what's wrong. I share my code below for reference:

```
import time

import cflib.crtp
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie

uris = [
    'radio://0/80/2M/E7E7E7E701',
    'radio://0/80/2M/E7E7E7E702',

    # Add more URIs if you want more copters in the swarm
]



def take_off(scf):
    
    commander= scf.cf.high_level_commander
    uri = scf.cf.link_uri
    if uri[-2:] == '01':
        commander.takeoff(0.5, 2)
        time.sleep(4)
        commander.go_to(0.0, 0.9,0.5,0,1)
        time.sleep(3)
    else:
        commander.takeoff(0.5, 2)
        time.sleep(4)
        commander.go_to(0.0,-0.9,0.5,0,1)
        time.sleep(3)
        
def land(scf):
    commander= scf.cf.high_level_commander

    commander.land(0.0, 2.0)
    time.sleep(2)

    commander.stop()

def hover_sequence(scf):
    take_off(scf)
    land(scf)

def reset_estimator(scf): # reset estimators not working properly
    
    cf = scf.cf
    uri = scf.cf.link_uri

    init_pos1 = [0.0, 1.0, 0.0]
    init_pos2 = [0.0,-1.0, 0.0]
    
    if uri[-2:] == '01':
    # Set Initial Values
        cf.param.set_value('kalman.initialX', str(init_pos1[0]))
        time.sleep(0.1)
        cf.param.set_value('kalman.initialY', str(init_pos1[1]))
        time.sleep(1)
    else:
        cf.param.set_value('kalman.initialX', str(init_pos2[0]))
        time.sleep(0.1)
        cf.param.set_value('kalman.initialY', str(init_pos2[1]))
        time.sleep(1)
    
    # Reset the Kalman Filter (Important!)
    cf.param.set_value('kalman.resetEstimation', '1')
    time.sleep(0.1)
    cf.param.set_value('kalman.resetEstimation', '0')
    time.sleep(2)

    scf.cf.param.set_value('stabilizer.controller','1')
    time.sleep(0.1)
    scf.cf.param.set_value('stabilizer.estimator', '2')
    time.sleep(1)



if __name__ == '__main__':
    cflib.crtp.init_drivers(enable_debug_driver=False)
    factory = CachedCfFactory(rw_cache='./cache')
    with Swarm(uris, factory=factory) as swarm:
        print('Connected to  Crazyflies')
        
        swarm.parallel(reset_estimator)

        time.sleep(2)

        swarm.parallel(hover_sequence)
```

Could anyone please help me out on understanding what's going wrong?

Thanksin advance.

**Comment by knmcguire** [comment]
Hi! So I see you have already found the step by step guide for swarming: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/user-guides/sbs_swarm_interface/

How does each individual crazyflie fly? Have you tried just connecting them to each of them through the cfclient and just have a quick test of take off with the command based flight control of the [flight control tab](https://www.bitcraze.io/documentation/repository/crazyflie-clients-python/master/userguides/userguide_client/flightcontrol_tab/)?

**Reply by Allanadn** [reply]
Hi Kimberly,

Thank you for your answer. Yes, I have tried to base my code on the step by step guide for swarming :) .

When running the above script, the crazyflie connected to URI: radio://0/80/2M/E7E7E7E701, just takes off and land. The crazyflie connected to URI radio://0/80/2M/E7E7E7E702, goes to the opposite expected direction and keeps flying until it hits the wall. Ex.: Starting from y=-1, I would expect it to go to y=-0.9, but it goes to a coordinate smaller than y=-1. 

I have connected through the cfclient and take off with them individually. I have also (in a modified version of the script) managed to make both take off and hover on different heights depending on the URI assigned for each UAV (although this task would have an erratic behavior from time to time). 

Not sure if it helps but I am using the crazyradio pa 1.0. Would it help to assign a single UAV to a single radio? If so, how could I use it?

**Reply by knmcguire** [reply]
Hi! 

The crazyradio version shouldn't matter in these case as they should use the same protocol. 


Hmm.. what I think that might be happening with the faulty crazyflie deck, is that while it is on the ground, the position converges because the flowdeck is too close to the ground or looking at something black. Therefore, it probably thinks it is somewhere completely different.

The take off button on the cfclient is relative from any postion that the crazyflie is on. If you look at the state estimate  of the positionof the faulty crazyflie with flowdeck, if you put it on the ground, do you see the position converge?

And if you remove the reset estimate and just have the crazyflies take off and land (comment out the to_go), what happens then?

**Reply by Allanadn** [reply]
I see... This is the output that I get from the estimator:

URI: radio://0/80/2M/E7E7E7E701 (supposedly healthy):

![image](https://github.com/bitcraze/discussions/assets/33231979/5637ae20-8214-4839-8fd8-a05908f0ebaf)

URI: radio://0/80/2M/E7E7E7E702 (the one having a strange behavior):

![image](https://github.com/bitcraze/discussions/assets/33231979/373f32b4-a447-4572-8109-f5a3a3883a1f)

When performing the steps you mentioned above, both of them take off, but none of them hover in place, and start moving to some other "random" position until they land...

**Comment by knmcguire** [comment]
Hi! 

I tried your code out with 2 crazyflies and it works alright with me! I've written something myself but should be pretty similar (see here underneath).

So I think that there is something either wrong with your flowdecks or the environment. This is the floor I flew on which works pretty fine, which is in a bright area with multiple sources of light. If your room only have one source of light right on top of the crazyflie, it could also be that it's moving on its shadow (tricking the optical flow measurements) which can cause similar issues. 

![20231121_124703](https://github.com/bitcraze/discussions/assets/11061260/d0645821-eac5-445e-a77b-ea08c635c7e2)

```
import time

import cflib.crtp
from cflib.crazyflie.swarm import CachedCfFactory
from cflib.crazyflie.swarm import Swarm
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie



def take_off(scf: SyncCrazyflie):
    commander = scf.cf.high_level_commander

    commander.takeoff(0.5, 2.0)
    uri = scf.cf.link_uri

    if uri[-2:] == '02':
        commander.takeoff(0.5, 2)
        time.sleep(4)
        commander.go_to(0.0, 0.7,0.5,0,1)
        time.sleep(3)
    else:
        commander.takeoff(0.5, 2)
        time.sleep(4)
        commander.go_to(0.0,-0.7,0.5,0,1)
        time.sleep(3)
    time.sleep(3)


def land(scf: SyncCrazyflie):
    commander = scf.cf.high_level_commander

    commander.land(0.0, 2.0)
    time.sleep(2)

    commander.stop()


def reset_estimator(scf): # reset estimators not working properly
    
    cf = scf.cf
    uri = scf.cf.link_uri

    init_pos1 = [0.0, 0.5, 0.0]
    init_pos2 = [0.0,-0.5, 0.0]
    
    if uri[-2:] == '02':
    # Set Initial Values
        cf.param.set_value('kalman.initialX', str(init_pos1[0]))
        time.sleep(0.1)
        cf.param.set_value('kalman.initialY', str(init_pos1[1]))
        time.sleep(1)
    else:
        cf.param.set_value('kalman.initialX', str(init_pos2[0]))
        time.sleep(0.1)
        cf.param.set_value('kalman.initialY', str(init_pos2[1]))
        time.sleep(1)
    
    # Reset the Kalman Filter (Important!)
    cf.param.set_value('kalman.resetEstimation', '1')
    time.sleep(0.1)
    cf.param.set_value('kalman.resetEstimation', '0')
    time.sleep(2)

    scf.cf.param.set_value('stabilizer.controller','1')
    time.sleep(0.1)
    scf.cf.param.set_value('stabilizer.estimator', '2')
    time.sleep(1)

uris = [
    'radio://0/40/2M/E7E7E7E707',
    'radio://0/40/2M/E7E7E7E702',
]


if __name__ == '__main__':
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')
    with Swarm(uris, factory=factory) as swarm:
        print('Connected to  Crazyflies')

        swarm.reset_estimators()
        print('Estimators have been reset')
        swarm.parallel_safe(reset_estimator)
        swarm.parallel_safe(take_off)
        swarm.parallel_safe(land)
```

**Reply by Allanadn** [reply]
Hi Kimberly, thank you so much for testing out my code and following up on that.

I'll test your code, but I do have only one source of light above, in the ceiling and the floor does not have an uneven pattern like yours. Does any other bitcraze sensor perform better under these conditions?

Thank you!

**Reply by knmcguire** [reply]
ah yeah the one light source on the top, if it is from a single spot (not a tube light), that might cause problems as I seen it before. If it casts a shadow underneath the crazyflie, it will move 'with the shadow' causing all kinds of problems. 

If you want to do 'onboard sensor only' positioning there no other sensor that we have on shelve that will work better. You have to make sure that there is good texture on the floor (you can use paint tape for that), and there is good lightning (perhaps use a desk light to shine from another side. If you'd like better positioning, then you should use  an external positionning system like MoCap, Lighthouse or the LPS system (see here: https://www.bitcraze.io/documentation/system/positioning/ or this blogpost https://www.bitcraze.io/2021/05/positioning-system-overview/)

**Reply by Allanadn** [reply]
Oh, okay this is good to know, thanks! On the ceiling it is a tube light and for patterns on the floor I tried to put some colored objects around, nonetheless, the problem still persists. Will try to maybe include some other light sources.

**Reply by knmcguire** [reply]
It is very important that the drone also starts from an area that it can see texture as well (perhaps more small and refined like the wood floor). In the cfclient you can see if the position diverges when it is placed on an fully plain color surface, and perhaps you can implement a check for that in your script (some of the autonomy scripts have a good check for that). You can also check the optical flow measurements by observing the motion  log parameter, and just holding it in your hand and move it manually over the surface. Also try it out on different areas as well to confirm.

At least you know that it's not a problem in your script.

**Reply by Allanadn** [reply]
Hi Kimberly,

Sorry for the long time to come bck to you on this. I believe you were right. The flowdeck 2 of the crazyflie behaving strangely is completely off. Here is a screenshot of the log I took when it was standing still on a surface like your wooden floor:

![image](https://github.com/bitcraze/discussions/assets/33231979/0f507b33-f8d6-4833-80ce-321b0e158d48)

When changing to another flowdeck at the same surface this is what I get:

![image](https://github.com/bitcraze/discussions/assets/33231979/5605e7d8-a73b-485c-afc5-ea2851ec65aa)

Thank you so much for the help!

**Reply by knmcguire** [reply]
ah great! then perhaps that was a bad flow-deck... if you just bought the flowdeck from us then we should probably send a replacement. So send us and email at contact _AT_ bitcraze.io with the order number

**Reply by Allanadn** [reply]
Thank you so much! Unfortunately this is a quite old flowdeck, so I bet is just time that took its toll on it :). Btw what's the role of scf: SyncCrazyflie for instance in : "def take_off(scf: SyncCrazyflie):" is there any difference in assigning just def take_off(scf): ?

**Reply by knmcguire** [reply]
Ah yeah that is a bit of inconsistency I guess in the codem, but this is more of a python question.
`scf:SyncCrazyflie` is a type hint, that the function should take the input of the class SyncCrazyflie and nothing else. Practically there is no differences for you though.
